src/Path/Include.hs

```diff
@@ -334,47 +346,105 @@ dirname (Path "") = Path ""
+-- >>> fileExtension $(mkRelFile "name.."       ) == ""
+-- >>> fileExtension $(mkRelFile ".name"        ) == ""
+-- >>> fileExtension $(mkRelFile "..name"       ) == ""
+--
 -- @since 0.5.11
 fileExtension :: Path b File -> String
-fileExtension = FilePath.takeExtension . toFilePath
-
--- | Add extension to given file path. Throws if the
--- resulting filename does not parse.
---
--- >>> addFileExtension "txt $(mkRelFile "foo")
--- "foo.txt"
--- >>> addFileExtension "symbols" $(mkRelFile "Data.List")
--- "Data.List.symbols"
--- >>> addFileExtension ".symbols" $(mkRelFile "Data.List")
--- "Data.List.symbols"
--- >>> addFileExtension "symbols" $(mkRelFile "Data.List.")
--- "Data.List..symbols"
--- >>> addFileExtension ".symbols" $(mkRelFile "Data.List.")
--- "Data.List..symbols"
--- >>> addFileExtension "evil/" $(mkRelFile "Data.List")
--- *** Exception: InvalidRelFile "Data.List.evil/"
+fileExtension = snd . splitExtension
+
+-- | Add extension to given file path.
+--
+-- >>> addFileExtension ".foo"   $(mkRelFile "name"     ) == "name.foo"
+-- >>> addFileExtension ".foo."  $(mkRelFile "name"     ) == "name.foo."
+-- >>> addFileExtension ".foo.." $(mkRelFile "name"     ) == "name.foo.."
+-- >>> addFileExtension ".bar"   $(mkRelFile "name.foo" ) == "name.foo.bar"
+--
+-- Throws an 'InvalidExtension' exception if the extension is not valid. A
+-- valid extension starts with a @.@ followed by one or more characters not
+-- including @.@ followed by zero or more @.@s in trailing position. Moreover,
+-- an extension must be a valid filename, notably it cannot include path
+-- separators. Particularly, @.foo.bar@ is an invalid extension, instead you
+-- have to first set @.foo@ and then @.bar@ individually.
+--
+-- The following law holds:
+--
+-- @(fileExtension . fromJust . addFileExtension ext) f == ext@
 --
 -- @since 0.6.1
 addFileExtension :: MonadThrow m
   => String            -- ^ Extension to add
   -> Path b File       -- ^ Old file name
   -> m (Path b File)   -- ^ New file name with the desired extension added at the end
-addFileExtension ext (Path path) =
-  if FilePath.isAbsolute path
-    then liftM coercePath (parseAbsFile (FilePath.addExtension path ext))
-    else liftM coercePath (parseRelFile (FilePath.addExtension path ext))
-  where coercePath :: Path a b -> Path a' b'
-        coercePath (Path a) = Path a
+addFileExtension ext (Path path) = do
+    validateExtension ext
+    return $ Path (path ++ ext)
 
--- | A synonym for 'addFileExtension' in the form of an operator.
--- See more examples there.
---
--- >>> $(mkRelFile "Data.List") <.> "symbols"
--- "Data.List.symbols"
--- >>> $(mkRelFile "Data.List") <.> "evil/"
--- *** Exception: InvalidRelFile "Data.List.evil/"
+    where
+
+    validateExtension ex@(sep:xs) = do
+        -- has to start with a "."
+        when (not $ FilePath.isExtSeparator sep) $
+            throwM $ InvalidExtension ex
+
+        -- just a "." is not a valid extension
+        when (xs == []) $
```

Comment:
Its just a matter of taste I guess. Here we are just pattern matching on a constructor so the code is as efficient and simple as it can be. But yes, `null` can be used if we want to follow a general convention across containers.